include/rapidcheck/state/Commands.hpp

```diff
@@ -49,5 +196,81 @@ void showValue(const std::vector<
   }
 }
 
+template <typename CommandType>
+struct ParallelCommands {
+  using Cmd = CommandType;
+  using Cmds = Commands<CommandType>;
+
+  ParallelCommands() {}
+
+  ParallelCommands(const Cmds &prefix, const Cmds &left, const Cmds &right)
+      : prefix(prefix)
+      , left(left)
+      , right(right) {}
+
+  Cmds prefix;
+  Cmds left;
+  Cmds right;
+};
+
+template <typename Cmd, typename Model, typename Sut>
+void runAllParallel(const ParallelCommands<Cmd> &commands, const Model &state, Sut &sut) {
+
+  // Verify pre-conditions for all possible interleavings
+  detail::isValidSequence(commands, state);
+
+  detail::ParallelExecutionResult<Model, Cmd> result;
+
+  // Run serial commands
+  for (const auto &command : commands.prefix) {
+    auto verifyFunc = command->run(sut);
+    result.prefix.emplace_back(
+        detail::CommandResult<Model, Cmd>(command, std::move(verifyFunc)));
+  }
+
+  rc::detail::Barrier b(2);
+
+  // Run the two parallel command sequences in separate threads
+  auto parallelCommandSeq1 = commands.left;
+  auto t1 = std::thread([&parallelCommandSeq1, &result, &sut, &b] {
+    b.wait();
+    for (const auto &command : parallelCommandSeq1) {
+      auto verifyFunc = command->run(sut);
+      std::this_thread::yield();
+      result.left.emplace_back(
+          detail::CommandResult<Model, Cmd>(command, std::move(verifyFunc)));
+    }
+  });
+
+  auto parallelCommandSeq2 = commands.right;
+  auto t2 = std::thread([&parallelCommandSeq2, &result, &sut, &b] {
+    b.wait();
+    for (const auto &command : parallelCommandSeq2) {
+      auto verifyFunc = command->run(sut);
+      std::this_thread::yield();
+      result.right.emplace_back(
+          detail::CommandResult<Model, Cmd>(command, std::move(verifyFunc)));
+    }
+  });
+
+  t1.join();
```

Comment:
Either that or simply two local `exception_ptr`s. However, it might be a good idea to break out the execution thread thingie so as to not repeat ourselves. After all, we're doing the exact same thing twice here.

A `promise` could actually work but in that case I'd rather see that the execution result is the value type of the promise and that `.get()` either returns the result or throws the appropriate exception. In that case, it's probably nicer to use `std::async` instead of `std::thread` since that doesn't require you to both `join()` and `get()`. A `get()` will suffice.